Port Kodi to FFmpeg 5.0

basilgello opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
This PR is a first attempt to prepare Kodi for upcoming Debian transition to FFmpeg 5.0

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
There is a bug report filed by @sramacher about Kodi and inputstream.ffmpegdirect failing to build against FFmpeg 5,0 in Debian experimental.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Tested on Debian sid/bookworm locally and by Debian users.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Debian and other Linux users will be able to continue using Kodi on newer systems.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed
- [ ] 

Linked issues:
closes https://github.com/xbmc/xbmc/issues/21625
closes https://github.com/xbmc/xbmc/issues/21766

lrusak commented:
You should discuss this stuff in slack first as there is at least 3 people working on this already.

lrusak commented:
Also, FYI https://github.com/xbmc/xbmc/pull/16185

basilgello commented:
> You should discuss this stuff in slack first as there is at least 3 people working on this already.

I did this when everyone was at DevCon :) And I expected it was an easy move!

AlwinEsch committed:
FFmpeg5 port

basilgello committed:
Use Channel Layout API if built against FFmpeg 5.1

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello committed:
ffmpeg5: Get extradata with extract_extradata BSF

Fixes the transport stream playback failures described in
https://bugs.debian.org/1016925

Rogo95 made an excellent technical analysis of the root cause
and reported that to the bug thread.

Later on, James Almer (jamrial) suggested the solution to use
extract_extradata bitstream filter to replace the removed split()
function.

Finally, I adapted the following code snippet:
https://gist.github.com/moonpfe/f6795d51294d91ee0f82f62ff6985db0
to Kodi and tested it by playing the affected files in TS format.

HiassofT form LibreELEC found another edge case decoding HTSP
streams from pvr.hts. The comparison of log files revealed that
the split function is also used in DVDDemuxClient.

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello commented:
@AlwinEsch this fixes playback of TS files including the one you gave me in Slack (sky...)
@phunkyfish You asked about the draft PR

wsnipex commented:
please note that in it's current state, jenkins is using ffmpeg 4.4 for building this PR

phunkyfish commented:
@basilgello we’ll need to get a kodi branch for ffmpeg 5 created from our clone. @lrusak can you help with this?

Once that is there we need to update this file in your PR: https://github.com/xbmc/xbmc/blob/master/tools/depends/target/ffmpeg/FFMPEG-VERSION

Did you require any patches for the ffmpeg5 build?

basilgello commented:
> Did you require any patches for ffmpeg5 build?

If I understood your question correctly, I do not need any modifications to ffmpeg5 source tree to run this.
I linked against ffmpeg packaged in Debian, and it has no patches Kodi adds to ffmpeg git tree.

phunkyfish commented:
> > Did you require any patches for ffmpeg5 build?
> 
> If I understood your question correctly, I do not need any modifications to ffmpeg5 source tree to run this. I linked against ffmpeg packaged in Debian, and it has no patches Kodi adds to ffmpeg git tree.

Ok, great. Then it should be easy to create and ffmpeg 5 branch.

basilgello commented:
Yes, 5.1.2 is better because I made all the tests with it.

phunkyfish commented:
A general comment on this PR. There are quite a few `#if` macros added based on particular versions of ffmpeg components. 

I’m wondering if it’s possible to also add the ffmpeg release (maybe as a comment) so I’m the future it’s easier to tell when a macro could be removed.

Is there value in adding this or is everyone fine with just the component version checks?

basilgello commented:
In theory there is `ff_const59` but it is only in libavformat and it is marked as volatile.

a1rwulf commented:
Afaik we ever only supported one version of ffmpeg at the same time.
Why do we need those `#ifdefs` at all?

phunkyfish commented:
Because Linux users may not have ffmpeg 5, they may only have 4. And as we don’t build ffmpeg for Linux we can force it.

lrusak commented:
> Because Linux users may not have ffmpeg 5, they may only have 4. And as we don’t build ffmpeg for Linux we can’t force it.
> 
> (I think)

This isn't quite right, we detect the version we want, if not found we build the internal version. You can force building an external ffmpeg but we still require it to be compatible (think LibreELEC, etc).

lrusak commented:
@basilgello You can use the new tag version here -> https://github.com/xbmc/FFmpeg/releases/tag/5.1.2-Nexus-Alpha3

phunkyfish commented:
Oh, so we don’t really need the ifdefs at all. Just update the cmake script used on Linux?

phunkyfish commented:
> 

Weird, this is empty on your reply

lrusak commented:
> Oh, so we don’t really need the ifdefs at all. Just update the cmake script used on Linux?

Well, the depends build of ffmpeg uses the CMakeLists.txt found here -> https://github.com/xbmc/xbmc/blob/master/tools/depends/target/ffmpeg/CMakeLists.txt

This CMakeLists.txt is also used with using the cmake option `-DENABLE_INTERNAL_FFMPEG=ON`

There is a build script here https://github.com/xbmc/xbmc/blob/master/tools/depends/target/ffmpeg/autobuild.sh
I'm not sure if it's needed anymore though, I think it was to just configure an ffmpeg build manually (though it could be used for ppa builds, I'm not 100% sure).

wsnipex commented:
autobuild.sh is used for PPA packaging: https://github.com/xbmc/xbmc/blob/master/tools/Linux/packaging/package-kodi-launchpad.sh#L43 but it's only downloading the tarball.

basilgello commented:
I can try keeping ifdeffery in `cores/FFmpeg.{cpp,h}` and interface out the helper functions.

We also can go ffmpeg 5 fully for v20, and it is fine for me as Kodi from Debian maintainer. Any minimally tech-savvy Linux user intending to build Kodi 20 against FFmpeg 4 can revert a couple of patches or add the ifdeffery from intermediate versions of this PR.

Should I bother with helper functions approach now?

wsnipex commented:
traditionally, we only ever supported 1 ffmpeg major version, so going to 5.x for Nexus would be preferred if doable.
@fuzzard ?

basilgello commented:
OK, then I drop the ifdeffery with a separate commit and when it is merged and someone asks "How do I build Kodi 20 with ffmpeg 4", we will point to that specific commit.

basilgello committed:
Separate commit to remove FFmpeg 4 support

Team Kodi does officially support one major FFmpeg version.
This commit is intentionally separated to let *nix distro
maintainers revert it if building Kodi 20+ is needed with
FFmpeg 4.

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello committed:
Bump depends-build ffmpeg version to 5.1.2

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

a1rwulf commented on xbmc/cdrip/EncoderFFmpeg.cpp:
```diff
@@ -327,19 +313,19 @@ bool CEncoderFFmpeg::WriteFrame()
     }
   }
-  catch (EncoderException& caught)
+  catch (FFMpegException& caught)
```
Can this be `const FFMpegException& caught`?

a1rwulf commented on xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp:
```diff
@@ -242,63 +248,78 @@ unsigned int CAEEncoderFFmpeg::GetFrames()
-  if (!pkt)
+    size = pkt->size;
+  }
+  catch (FFMpegException& caught)
```
Can this be const FFMpegException& caught?

a1rwulf commented on xbmc/cores/AudioEngine/Interfaces/AEStream.h:
```diff
@@ -13,6 +13,7 @@
 
 extern "C" {
 #include <libavcodec/avcodec.h>
+#include <libavutil/channel_layout.h>
```
Is there any particular reason why you added it to the header?
Is it sufficient to put it in the cpp file where it's needed?

a1rwulf commented on xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodec.h:
```diff
@@ -16,6 +16,7 @@
 
 extern "C" {
 #include <libavcodec/avcodec.h>
+#include <libavutil/channel_layout.h>
```
Is this really needed in this header?

a1rwulf commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -647,7 +638,10 @@ bool CDVDDemuxFFmpeg::Open(const std::shared_ptr<CDVDInputStream>& pInput, bool
             AVStream* st = m_pFormatContext->streams[idx];
-            if ((st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO && st->codec_info_nb_frames > 0) ||
+            // Related to https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210429143825.53040-1-jamrial@gmail.com/
+            // was becomed replaced with AVSTREAM_EVENT_FLAG_NEW_PACKETS.
```
`was becomed` => `has been` maybe?

a1rwulf reviewed (commented):
I can't really say how correct the ffmpeg code is, so maybe someone else can have a look too.
Commented on some minors.
Thanks for working on this!

basilgello commented:
@a1rwulf Addressed the comments and fixed the remnants of cleanup breaking POSIX builds!

phunkyfish commented:
jenkins build this please

HiassofT commented:
I did a quick test with this PR on LibreELEC/RPi4 and while all files so far played fine I noticed that trying to play live TV with pvr.hts just gives a "waiting circle". Playing back recordings through pvr.hts worked fine though.

Current kodi master works fine here, I tested with this PR and LE/RPi ffmpeg both with software and hardware decoding and finally dropped all RPi and LE ffmpeg patches so I only used ffmpeg 5.1.2 plus the 5.1.2-Nexus-Alpha3 patches - still it hung with software decoding.

Here's a log file with ffmpeg component logging enabled, trying to play a SD H264 stream: http://ix.io/4dMh - note the repeating `debug <general>: ffmpeg[0x3613e90]: [h264] ct_type:0 pic_struct:3` messages

And here's a log trying to play a HD H265 stream: http://ix.io/4dMg

basilgello commented:
Thanks @HiassofT for the testing! May I also ask you to test audio encoding and audio channel remapping? In the next force push I will add another big commit  switching to channel layout APIs. And of course I will check the log you provided!

HiassofT commented:
I probably won't be able to help much with channel mapping/layout testing.

I don't have any of the modern multi-channel rubbish here, analog stereo NAD amplifier and 2.0 speakers FTW :-)

HiassofT commented:
I did a quick test on my laptop running Debian Bullseye and pvr.hts live TV playback hangs there, too, http://ix.io/4dPU

basilgello commented:
@HiassofT All your logs show `unsupported stream` error that shoukd have been fixed with extradata commit. However, debug log shows no traces of extradata being called at all. I recall TVHeadEnd streams use RTSP ? That might explain why extradata step is omitted.

HiassofT commented:
tvheadend uses it's own tcp based HTSP protocol https://tvheadend.org/projects/tvheadend/wiki/Htsp - but don't ask me any specifics about it, kodi pvr addon maintainers might help you more there

The "unsupported stream" warning occurs with ffmpeg 4.4, too.

For reference I uploaded logs with ffmpeg logging and pvr.hts tracing enabled, once with latest version of this ffmpeg 5.1.2 PR and once with 4.4.1

5.1.2: https://www.horus.com/~hias/tmp/libreelec/kodi-5.1.2.log
4.4.1: https://www.horus.com/~hias/tmp/libreelec/kodi-4.4.1.log

pvr.hts seems to relay the stream packets fine to kodi, but it looks like no H264 packet parsing is happening with 5.1.2

basilgello commented:
@lrusak can you please add a Kodi patch to ffmpeg changing iOS from 11 to 13 here https://github.com/xbmc/FFmpeg/blob/release/5.1-kodi/libavfilter/metal/utils.m#L34 and update the tarball ? This should fix iOS builds and make Jenkins evergreen :)

basilgello commented:
@HiassofT Good catch! The logs show that parsing takes place but without extradata split.

basilgello commented:
@HiassofT Please test the latest PR :)

HiassofT commented:
Thanks, this is better, live TV playback is working now - tested both with software decoding and drmprime / v4l2 HW decoding on LibreELEC/RPi4

HiassofT commented:
While extract_extradata seems to have helped getting live TV playback working kodi log is now spammed with ffmpeg errors during playback
```
2022-10-27 12:19:26.554 T:1743    error <general>: ffmpeg[0x29eea70]: [extract_extradata] Codec 'eac3' (86056) is not supported by the bitstream filter 'extract_extradata'. Supported codecs are: av1 (226) avs2 (192) avs3 (194) cavs (87) h264 (27) hevc (173) mpeg1video (1) mpeg2video (2) mpeg4 (12) vc1 (70) 
2022-10-27 12:19:29.395 T:1750     info <general>: Skipped 93 duplicate messages..
```
Full log is here: http://ix.io/4eeR

enen92 commented:
In case someone needs a testbuild I asked jenkins to build one with the py3 patch + the ffmpeg update for android (arm32) for personal use:
https://mirrors.kodi.tv/test-builds/android/arm/kodi-20221027-e62535fe-ffmpeg+py311-27074-armeabi-v7a.apk

Haven't find any regression/non-playable video yet.

basilgello commented:
@enen92 @HiassofT please test with this revision! It should fix spamminess in the log

lrusak commented:
> @lrusak can you please add a Kodi patch to ffmpeg changing iOS from 11 to 13 here https://github.com/xbmc/FFmpeg/blob/release/5.1-kodi/libavfilter/metal/utils.m#L34 and update the tarball ? This should fix iOS builds and make Jenkins evergreen :)

Do we actually need that? Can't we just disable the filter with `--disable-filter=yadif_videotoolbox`?

lrusak commented on xbmc/cores/FFmpeg.cpp:
```diff
@@ -117,3 +135,132 @@ void ff_avutil_log(void* ptr, int level, const char* format, va_list va)
   buffer.erase(0, start);
 }
 
+#define FF_MAX_EXTRADATA_SIZE ((1 << 28) - AV_INPUT_BUFFER_PADDING_SIZE)
```
constexpr in anonymous namespace

lrusak commented on xbmc/cores/FFmpeg.cpp:
```diff
@@ -117,3 +135,132 @@ void ff_avutil_log(void* ptr, int level, const char* format, va_list va)
+
+  *p_extradata = nullptr;
+
+  AVBSFContext* bsf = nullptr;
```
Please declare variables where they are used.

lrusak commented on xbmc/cores/FFmpeg.cpp:
```diff
@@ -117,3 +135,132 @@ void ff_avutil_log(void* ptr, int level, const char* format, va_list va)
+      *p_extradata = (uint8_t*)av_malloc(ret_extradata_size + AV_INPUT_BUFFER_PADDING_SIZE);
+      if (!*p_extradata)
+      {
+        CLog::Log(LOGERROR, "{} - failed to allocate {} bytes for extradata", __FUNCTION__,
```
`CLog::LogF`

lrusak commented on xbmc/cores/FFmpeg.cpp:
```diff
@@ -117,3 +135,132 @@ void ff_avutil_log(void* ptr, int level, const char* format, va_list va)
+        return 0;
+      }
+
+      CLog::Log(LOGDEBUG, "{} - fetching extradata, extradata_size({})", __FUNCTION__,
```
`CLog::LogF`

lrusak commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp:
```diff
@@ -149,41 +148,44 @@ bool CDVDDemuxClient::ParsePacket(DemuxPacket* pkt)
     {
-      if (st->ExtraData)
-        delete[] st->ExtraData;
+      CLog::Log(LOGERROR, "CDVDDemuxClient::{} - av_packet_alloc failed: {}", __FUNCTION__,
```
`CLog::LogF`

lrusak commented on xbmc/cores/FFmpeg.h:
```diff
@@ -21,6 +22,28 @@ extern "C" {
 }
 
+// https://github.com/FFmpeg/FFmpeg/blob/56450a0ee4/doc/APIchanges#L18-L26
+#define FFMPEG_FMT_CONST const
```
Can we get rid of this define and just use `const`?

lrusak commented on xbmc/cores/FFmpeg.h:
```diff
@@ -51,3 +74,7 @@ class CFFmpegLog
   int level;
 };
 
+int GetPacketExtradata(const AVPacket* pkt,
```
You could return `uint8_t*` (which is the allocated extradata), That way you can avoid the double pointer. You can use the address for the size `int& extraDataSize` or just return a `std::tuple<uint8_t*, int>`.

lrusak reviewed (commented):
Just some minors and formatting issues.

Please be aware of the code guidelines for variable names -> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-variables

basilgello committed:
Disable broken filter for darwin_embedded

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello commented:
@lrusak I adderessed the issues you spotted, thanks! Also, disabling the filter you mentioned made ios-arm64 target green :)

@enen92 @HiassofT please re-test the latest rev.

Tomorrow I will drop of the net for one week (maybe even more) so if there are no serious issues with the review, let's merge it!

CastagnaIT commented on xbmc/cores/FFmpeg.cpp:
```diff
@@ -117,3 +135,130 @@ void ff_avutil_log(void* ptr, int level, const char* format, va_list va)
+  AVPacket* dstPkt = av_packet_alloc();
+  if (!dstPkt)
+  {
+    CLog::LogF(LOGERROR, "{} - failed to allocate packet", __FUNCTION__);
```
when you use `CLog::LogF` the `__FUNCTION__` is not needed because already implicit
so become as
```suggestion
    CLog::LogF(LOGERROR, "Failed to allocate packet");
```
you need to fix also all others

basilgello replied:
Done, thanks!

thexai committed:
Updates FFmpeg Windows patches

thexai commented:
Runtime tested on Windows x64 and no (evident) issues found. I will continue using it for the next few days.

Windows patches in `tools/buildsteps/windows/patches` need to be updated to apply clean.

@basilgello, please add these patches changes: https://github.com/thexai/xbmc/commit/7aac790b13769fbed01b85b476aa73bcfc411bcd

basilgello commented:
@thexai Done, thanks!

HiassofT commented:
I finally got some time to do more extensive tests on LibreELEC/RPi4 and everything seems to be working fine here :+1:

phunkyfish commented:
Jenkins build this please

mriedel87 commented:
I also can confirm that these patches are working fine. 
Tested on Debian (Sid; arch x64) with VNSI as PVR client plugin.
Thank you all :)

thexai commented:
Tested also on Nvidia Shield 2019 with quietvoid's DoVi MKV patch (https://github.com/thexai/xbmc/commits/shield)

Tested some 4K MKV's that before to only play in HDR10 now play in DV without changing any Shield settings. All seems right 😀

KLudwig-V commented:
@fuzzard 

FFMPEG 5 was released January 2022... (5.1 in July). If this gets moved to v21 that means (statistically) one year from now, when the release is already 2 years old...

What is the risk of adding this before the first Beta? It seems it has been tested and works OK...

jjd-uk commented:
Yes it's tested, but only by a handful of people. With it being such a key Kodi component the felling was that we'd need to do at least another couple of alpha's to make sure the multitude of platforms Kodi runs on are adequately tested, and this could potentially severely delay v20. Once we get Beta done, then we can branch v20 from Master, then Master becomes v21 development so this could then be merged into Master and available in Nightlies.

jjd-uk commented:
However it's also recognised there's a big push from some quarters to get ffmpeg 5 in, so rather than delaying v20, the current thinking is to do a quick v21 before end of 2023. Not something we typically do, but there's nothing to stop us doing v21 only a few months after v20.

DecalicatanDecalicatan commented:
> However it's also recognised there's a big push from some quarters to get ffmpeg 5 in, so rather than delaying v20, the current thinking is to do a quick v21 before end of 2023. Not something we typically do, but there's nothing to stop us doing v21 only a few months after v20.

I guess you meant " to do a quick v21 before end of _**2022**_" ?

neo1973 commented on xbmc/cores/FFmpeg.cpp:
```diff
@@ -21,6 +21,19 @@
+  std::string text;
+  text.reserve(AV_ERROR_MAX_STRING_SIZE);
+  av_strerror(err, text.data(), AV_ERROR_MAX_STRING_SIZE);
+  return text;
```
I know this code already existed before but anyway it doesn't work because the `std::string` length is still zero: https://godbolt.org/z/P3eKTfWGz
I think the only way to fix this is to allocate a `char[AV_ERROR_MAX_STRING_SIZE]` on the stack and assign the `std::string` from there.

lrusak replied:
I think the solution is to use `resize` rather than `reserve`.

https://en.cppreference.com/w/cpp/string/basic_string/resize

thexai replied:
`AV_ERROR_MAX_STRING_SIZE` is only 64 so I think it's ok:

```c++
  text.resize(AV_ERROR_MAX_STRING_SIZE);
  av_strerror(err, text.data(), text.size());
```

If it were a much larger number (1000, 2000) then it would be better to do what neo1973 says to avoid always returning a string of maximum length.

neo1973 commented on xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp:
```diff
@@ -242,63 +261,80 @@ unsigned int CAEEncoderFFmpeg::GetFrames()
+    /* encode it */
+    err = avcodec_send_frame(m_CodecCtx, frame);
+    if (err < 0)
+      throw FFMpegException("Error sending a frame for encoding (error '{}')", __func__,
+                            FFMpegErrorToString(err));
```
The number of arguments doesn't match the format string. I think the `__func__` is a copy past mistake. Same a couple of lines below.

thexai commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp:
```diff
@@ -149,41 +149,44 @@ bool CDVDDemuxClient::ParsePacket(DemuxPacket* pkt)
+    AVPacket* avpkt = av_packet_alloc();
+    if (!avpkt)
+    {
+      CLog::LogF(LOGERROR, "CDVDDemuxClient::{} - av_packet_alloc failed: {}", __FUNCTION__,
```
```suggestion
      CLog::LogF(LOGERROR, "av_packet_alloc failed: {}", strerror(errno));
```

The` __FUNCTION__` parameter is not needed since `LogF` is used

HiassofT commented:
@basilgello a small request: could you please drop the "@" mention to me in the "ffmpeg5: Get extradata with extract_extradata BSF" commit message? Changing it to HiassofT, without the "@" or to my real name, Matthias Reichl, would be fine - in general I appreciate acknowledgements.

I'm getting rather annoying  (email) notifications whenever that commit is added to a branch on some github repo our when you rebase and force-push your PR branch and I haven't yet managed to configure github to stop doing that and it's filling my inbox...

basilgello commented:
@HiassofT Will do! Sorry!!! I thought fame would not be so annoying :)

phunkyfish commented:
Jenkins build this please

phunkyfish commented:
Jenkins build this please

phunkyfish commented:
Jenkins build this please

basilgello commented:
@garbear you reviewed #22309 which fixed a leak in extradata. I had to rebase this PR to handle it. Can you please have a look at the extradata commit, especially `std::unique_ptr<uint8_t[]>(retExtraData)`?

I made a reproducer and built it with `-fsanitize=leak` and no leaks were spotted, but I am not sure if this is valid approach re dangling pointers.

garbear commented:
Can you switch to `std::make_unique`? I don't see any problems in the meantime.

CastagnaIT commented on xbmc/cores/FFmpeg.cpp:
```diff
@@ -117,3 +135,128 @@ void ff_avutil_log(void* ptr, int level, const char* format, va_list va)
+        av_packet_get_side_data(pktRef, AV_PKT_DATA_NEW_EXTRADATA, &retExtraDataSize);
+    if (retExtraData && retExtraDataSize > 0 && retExtraDataSize < FF_MAX_EXTRADATA_SIZE)
+    {
+      extraData = (uint8_t*)av_malloc(retExtraDataSize + AV_INPUT_BUFFER_PADDING_SIZE);
```
use C++ cast

basilgello replied:
Thanks, I am testing the variant with `std::make_unique<uint8_t[]>` in this function!

CastagnaIT replied:
ah ok then nvm

basilgello replied:
The variant I tested mixes `operator new` and `av_freep` and thus crashes. At least I tested the assumption that ffmpeg itself does not free anything :)

basilgello replied:
nvm found an offending `av_freep` :)

neo1973 commented on xbmc/cores/FFmpeg.h:
```diff
@@ -21,6 +22,27 @@ extern "C" {
+    : s(StringUtils::Format(fmt, std::forward<Args>(args)...))
+  {
+  }
+  ~FFMpegException() throw() {} // Updated
+  const char* what() const throw() { return s.c_str(); }
```
`throw()` is deprecated/removed from C++, instead this should be:
```suggestion
  const char* what() const noexcept override { return s.c_str(); }
```

basilgello replied:
Fixed!

basilgello commented:
Should be ready for review and merge! Let's go?

garbear reviewed (approved):
Looks good from my end!

enen92 reviewed (approved):
I've been using this for quite a while so +1 from me
